proposals/0000-std-experimental.rst

```diff
@@ -0,0 +1,177 @@
+
+Effect and Interactions
+-----------------------
+The introduction of this flag provides a clear distinction between stable and
+experimental features. It ensures that developers are aware of the potential
+risks of using experimental features while providing a sandbox for testing out
+new GHC capabilities.
+
+Costs and Drawbacks
+-------------------
+- There might be an initial learning curve for developers to understand the
+  distinction between experimental and stable features.
+- Maintaining two sets of features (stable and experimental) might increase the
+  complexity of the compiler codebase.
+
+The authors believe the benefit of clearly separating experimental and those
+that adhere to a more rigourous change evolution easily offsets the costs for
+both end users and feature developers. For developers of experimental feature
+this also means they reserve the rigth to break them without warning in
+backwards incompatible ways at any moment in time leading to significantly
+lighter processes for experimental features.
+
+Prior Work
+----------
+The ``-experimental`` feature is similar to solutions for experimental
+features and implementation in other languages, and GHC would not have a
+bespoke solution here, but share a common approach with many other compilers.
+Do note that the provided approaches mix editions, channels, and feature gates.
+The goal is to illustrate that the concept of separating experimental features
+explicitly behind opt-in gates.
+
+- **Rust**: Rust has so called channels, which delineate stability by stable,
+  beta, and nightly. To opt-in locally to nightly features you'd use the
+  ``+nightly`` toolchain modifier. Effectively using a different Rust compiler
+  provided by ``rustup``.
+- **Java**: Java has the ``--enable-preview`` flag since Java 9, which allows
+  opt-in to experimental features.
+- **Gnu Compiler Collection**: GCC provides flags to opt-in to experimental
+  features through the ``-std=...`` flag. For example, it has flags like
+  ``-std=c++1z`` for experimental C++ features before they were standardized in
+  C++17.
+
+Backward Compatibility
+----------------------
+The introduction of this flag is expected to have minimal impact on existing
+code. The initial set of features behind the ``-experimental`` flag will be
+the emtpy set. Follow up proposals will be required to move existing features
+behind the `-experimental` flag. New features should by default be behind
+the `-experimental` flag, unless the respective authors consider them stable
+and not subject to sudden change.
+
+Conditionality Compilation
+--------------------------
+For maintainers there will be a ``__GHC_EXPERIMENTAL__`` macro to use with ``CPP``
+to allow for conditional compilation if ``-experimental`` is active. This can
+be used with the usual version macros to tailor to specific GHC versions as needed.
```

Comment:
What would you use this for?

Reply:
Let's assume you want to try some experimental feature but make it opt-in. You could have a flag in your cabal file `-fwith-experimental`, that then switches on `ghc-options: -experimental`, and you can CPP guard the experimental use in your module being a CPP guard.

Reply:
If you have a flag then you can simply pass your own CPP defines, so I don't see there being a reason for having this hard coded into the compiler. The only situation where you might need this is if you don't define a flag and expect users to add ghc-options in their cabal.project file but that doesn't seem like an idiomatic way to do this.

I seriously doubt as well that users will even do the flag approach as most of our tooling doesn't support optional features of packages. It would be better to just have two packages one that is experimental and one that isn't.

Reply:
Indeed you're not supposed to change API surface based on flags or CPPs. But that's not what the proposal implies.